### PR TITLE
Fixed segmentation fault in deflate_quick() when switching levels using deflateParam

### PR DESCRIPTION
**Pull request by nmoinvaz**
This is related to discussion on #536. 

I have added a unit test case that reproduces the segmentation fault. Here is the initial check-in of just the unit test causing the segmentation fault https://github.com/nmoinvaz/zlib-ng/runs/467173544 and afterwards after the fix is applied https://github.com/nmoinvaz/zlib-ng/runs/467198265.

Background:

- deflateInit would be initialized with a window size greater than 8K
- deflateParams called to switch to level 1 which does not update w_size
- Fault would occur because deflate_quick was not checking w_size bounds on dist when accessing quick_dist_codes

### CODE REVIEW

File: arch/x86/deflate_quick.c [hunk 1]
```diff
@@ -122,7 +124,7 @@ static inline long compare258(const unsigned char *const src0, const unsigned ch
 }
 
 static const unsigned quick_len_codes[MAX_MATCH-MIN_MATCH+1];
-static const unsigned quick_dist_codes[8192];
+static const unsigned quick_dist_codes[QUICK_WINDOW_SIZE];
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c]: Completely unnecessary change...

File: arch/x86/deflate_quick.c [hunk 2]
```diff
@@ -237,7 +243,7 @@ ZLIB_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
             hash_head = quick_insert_string(s, s->strstart);
             dist = s->strstart - hash_head;
 
-            if (dist > 0 && (dist-1) < (s->w_size - 1)) {
+            if (dist > 0 && (dist-1) < wsize) {
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c]: Should be ```wsize - 1```...

File: arch/x86/deflate_quick.c [hunk 3]
```diff
@@ -343,7 +349,7 @@ static const unsigned quick_len_codes[MAX_MATCH-MIN_MATCH+1] = {
     0x001c230d, 0x001d230d, 0x001e230d, 0x0000a308,
 };
 
-static const unsigned quick_dist_codes[8192] = {
+static const unsigned quick_dist_codes[QUICK_WINDOW_SIZE] = {
```
Comment (mtl1979) [inline arch/x86/deflate_quick.c]: Completely unnecessary change...

File: arch/x86/deflate_quick.c (hunk 2 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: I think here it should be `wsize`. I don't see the need for `-1` .  It is already a less than check.

File: arch/x86/deflate_quick.c (hunk 1 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: It seems proper that `quick_len_codes` depends on #define value so `quick_dist_codes` should as well. I had to create the #define for checking bounds, might as well use it where can.

File: arch/x86/deflate_quick.c (hunk 2 above)
Comment (mtl1979) [reply arch/x86/deflate_quick.c]: Obviously both matches need to fit in the window at the same time, otherwise the hash could not be calculated. The test is incorrect, but removing ``` - 1``` makes it less correct... Obviously ```dist``` can't be more than ```wsize - MIN_MATCH```.

File: arch/x86/deflate_quick.c (hunk 1 above)
Comment (mtl1979) [reply arch/x86/deflate_quick.c]: Because the array size doesn't change ever, using macro for the array size makes it prone for overflow.

File: arch/x86/deflate_quick.c (hunk 2 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: But then why wasn't the original check `s->w_size - 3`?

Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: > In the worst case the original string starts at byte index 0, and match can be last 3 bytes of window (8189, 8190, 8191). I don't know why it isn't w_size - 3, that is why I said the original test is also incorrect...

I think we should only care that `dist-1` is less than 8192  or `quick_dist_codes` array size, because in `static_emit_ptr`:
```
    unsigned code2 = quick_dist_codes[dist-1] >> 8;
    unsigned len2  = quick_dist_codes[dist-1] & 0xFF;
```

Comment (mtl1979) [reply arch/x86/deflate_quick.c]: I think we should add a comment here why it uses ```dist - 1``` in the second test... That way people don't need to scan the whole source to see the reason...

Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: I think I found the reason which is that `hash_mask` is window size -1.

**Comment by nmoinvaz** [comment]
This PR is unnecessary if #549 is accepted. But the test case can at least be salvaged.

File: arch/x86/deflate_quick.c (hunk 1 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: Ok I changed it for less code lines being changed.

File: arch/x86/deflate_quick.c (hunk 2 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: I have kept the -1.

File: arch/x86/deflate_quick.c (hunk 3 above)
Comment (nmoinvaz) [reply arch/x86/deflate_quick.c]: Changed.

**Comment by nmoinvaz** [comment]
Simplified and rebased.
